File: typescript/triangle/triangle.ts
export default class Triangle {
  sides: number[];

  constructor(...sides: number[]) {
    this.sides = sides;
  }

  kind() {
    let positiveSides = this.sides.filter((side: number) => side <= 0);
    if (positiveSides.length > 0) {
      throw "all sides must be positive";
    }

    let sidesBySize = this.sides.sort(
      (side1: number, side2: number) => side1 - side2
    );
    if (sidesBySize[2] > sidesBySize[1] + sidesBySize[0]) {
      throw "sides do not describe a triangle";
    }

    let typeByUniqueCount = ["equilateral", "isosceles", "scalene"];
    let uniqueCount = new Set(this.sides).size - 1;

    return typeByUniqueCount[uniqueCount];
  }
}

File: lua/triangle/triangle.lua
local triangle = {}

function triangle.kind(a, b, c)
    assert(a > 0, "Input Error")
    assert(b > 0, "Input Error")
    assert(c > 0, "Input Error")

    local sides = {a, b, c}
    table.sort(sides)

    if (sides[3] > sides[2] + sides[1]) then
        error('Input Error')
    end

    if (a == b and b == c) then
        return 'equilateral'
    end
    if (sides[1] == sides[2] or sides[2] == sides[3]) then
        return 'isosceles'
    end

    return 'scalene'
end

return triangle
